src/core/stream/orchestrator/stream_orchestrator.ts

```diff
@@ -421,6 +423,30 @@ export default function StreamOrchestrator(
         } else if (type === "active-stream") {
           // current Stream is active, destroy next Stream if created
           destroyNextStreams$.next();
+        } else if (
+          evt.type === "adaptationChange" &&
+          evt.value.type === "audio" &&
+          !evt.value.isFirstAdaptation
+          ) {
+          isReadyToAdaptAudioSwitchingStategy = true;
+        } else if (
+          evt.type === "added-segment" &&
+          evt.value.content.adaptation.type === "audio" &&
+          isReadyToAdaptAudioSwitchingStategy
+          ) {
+           const { audioTrackSwitchingMode } = options;
+           isReadyToAdaptAudioSwitchingStategy = false;
+          if (audioTrackSwitchingMode === "reload") {
+            return clock$.pipe(mergeMap((tick) => {
+                return observableMerge(
+                  observableOf(evt),
+                  observableOf(EVENTS.needsMediaSourceReload(basePeriod, tick)));
+            }));
+          } else  {
+              return observableMerge(
+                observableOf(evt),
+                observableOf(EVENTS.addedSegmentOnAdaptationChange("audio")));
```

Comment:
This should work and I'm relatively fine with it, but I still find it a little convoluted.

From my point of view here, we send more event information in a `PeriodStream` so we can send other events in the `StreamOrchestrator` so they can finally be handled by the `Init`.

Why not doing that code in the `PeriodStream` directly? The code would be split-up in less places and the `PeriodStream` would be more powerful in the possible future cases where we want  to use it as a stand-alone module (without a `StreamOrchestator`).

There is also already some code in the `PeriodStream` for adaptation switch strategies. We already did some work for example on video adaptation where we reload the media-source under some specific conditions.

Reply:
Hoo ok!

I understood, with more and more iteration, I get used to that part of the code and so, see the things.

By mixing both your recommendations, I think I got something cleaner:
- The main logic happens directly in the `period_stream` now.
  - I don't wait any pushed segment to fire the `needsMediaSourceReload` (didn't think of that before, but now you mention it, seems obvious :) ),
  - Then, I don't fire an `addedSegmentOnAdaptationChange` anymore but a `needsSourceBufferFlush` (recommended by @grenault73 ) event that is caught by the `init` to perform any action. Not sure about the name though.